embassy-stm32/src/usart/rx_ringbuffered.rs

```diff
@@ -0,0 +1,296 @@
+                w.set_idleie(false);
+            });
+            r.cr3().modify(|w| {
+                // disable Error Interrupt: (Frame error, Noise error, Overrun error)
+                w.set_eie(false);
+                // disable DMA Rx Request
+                w.set_dmar(false);
+            });
+        }
+    }
+
+    /// Read bytes that are readily available in the ring buffer.
+    /// If no bytes are currently available in the buffer the call waits until data are received.
+    ///
+    /// Background receive is started if `start()` has not been previously called.
+    ///
+    /// Receive in the background is terminated if an error is returned.
+    /// It must then manually be started again by calling `start()` or by re-calling `read()`.
+    pub async fn read(&mut self, buf: &mut [u8]) -> Result<usize, Error> {
+        let r = T::regs();
+
+        // SAFETY: read only
+        let is_started = unsafe { r.cr3().read().dmar() };
+
+        // Start background receive if it was not already started
+        if !is_started {
+            self.start()?;
+        }
+
+        // SAFETY: read only and we only use Rx related flags
+        let sr = unsafe { sr(r).read() };
+
+        let has_errors = sr.pe() || sr.fe() || sr.ne() || sr.ore();
+        if has_errors {
+            Self::teardown_uart();
+
+            if sr.pe() {
+                return Err(Error::Parity);
+            }
+            if sr.fe() {
+                return Err(Error::Framing);
+            }
+            if sr.ne() {
+                return Err(Error::Noise);
+            }
+            if sr.ore() {
+                return Err(Error::Overrun);
+            }
+        }
+
+        let ndtr = self.uart.ndtr();
+        self.ring_buf.ndtr = ndtr;
+        match self.ring_buf.read(&mut self.uart, buf) {
+            Ok(len) if len == 0 => {}
+            Ok(len) => {
+                assert!(len > 0);
+                return Ok(len);
+            }
+            Err(OverrunError) => {
+                // Stop any transfer from now on
+                // The user must re-start to receive any more data
+                Self::teardown_uart();
+                return Err(Error::Overrun);
+            }
+        }
+
+        // Wait for any data since `ndtr`
+        self.wait_for_data(ndtr).await?;
+
+        // ndtr is now different than the value provided to `wait_for_data()`
+        // Re-sample ndtr now when it has changed.
+        self.ring_buf.ndtr = self.uart.ndtr();
+        let len = self.ring_buf.read(&mut self.uart, buf).map_err(|_err| Error::Overrun)?;
+        assert!(len > 0);
+        Ok(len)
+    }
+
+    /// Wait for uart data
+    async fn wait_for_data(&mut self, old_ndtr: usize) -> Result<(), Error> {
+        let r = T::regs();
+
+        // make sure USART state is restored to neutral state when this future is dropped
+        let _drop = OnDrop::new(move || {
+            // SAFETY: only clears Rx related flags
```

Comment:
This is unsafe, but it seems that the problem is also present in the original UART driver. It doesn't matter if you don't touch TX flags, the read-modify-write (RMW) operation is performed on the whole register and can be interrupted by UART IRQ, which also uses some of these registers. This can also happen if TX/RX parts are used in different executors.

Reply:
Is there something in embassy that allows this to be done atomically with retry? Drone OS does this: https://github.com/drone-os/drone-core/blob/master/src/reg/atomic.rs#L88-L125